docker/runner.go

```diff
@@ -30,122 +45,201 @@ func NewRunner() *Runner {
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "unable to extract port from allocated listener address '%v'. %v", lis.Addr(), err)
+		return DefaultPort
+	}
+	return port
 
-	waitBodyChan, waitErrChan := cli.ContainerWait(ctx, cont.ID, container.WaitConditionNextExit)
+}
 
-	err = cli.ContainerStart(ctx, cont.ID, types.ContainerStartOptions{})
+func newContainer(ctx context.Context, c client.CommonAPIClient, f fn.Function, port string, verbose bool) (id string, err error) {
+	var (
+		containerCfg container.Config
+		hostCfg      container.HostConfig
+	)
+	if containerCfg, err = newContainerConfig(f, port, verbose); err != nil {
+		return
+	}
+	if hostCfg, err = newHostConfig(port); err != nil {
+		return
+	}
+	t, err := c.ContainerCreate(ctx, &containerCfg, &hostCfg, nil, nil, "")
 	if err != nil {
-		return errors.Wrap(err, "failed to start container")
+		return
 	}
-	defer func() {
-		t := time.Second * 10
-		err := cli.ContainerStop(context.Background(), cont.ID, &t)
-		if err != nil {
-			fmt.Fprintf(os.Stderr, "failed to stop container: %v", err)
-		}
-	}()
+	return t.ID, nil
+}
 
-	select {
-	case body := <-waitBodyChan:
-		if body.StatusCode != 0 {
-			return fmt.Errorf("failed with status code: %d", body.StatusCode)
-		}
-	case err := <-waitErrChan:
-		return err
-	case err := <-copyErrChan:
-		return err
-	case <-ctx.Done():
-		return ctx.Err()
+func newContainerConfig(f fn.Function, _ string, verbose bool) (c container.Config, err error) {
+	envs, err := newEnvironmentVariables(f, verbose)
+	if err != nil {
+		return
 	}
+	// httpPort := nat.Port(fmt.Sprintf("%v/tcp", port))
+	httpPort := nat.Port("8080/tcp")
+	return container.Config{
+		Image:        f.Image,
+		Env:          envs,
+		Tty:          false,
+		AttachStderr: true,
+		AttachStdout: true,
+		AttachStdin:  false,
+		ExposedPorts: map[nat.Port]struct{}{httpPort: {}},
+	}, nil
+}
 
-	return nil
+func newHostConfig(port string) (c container.HostConfig, err error) {
+	// httpPort := nat.Port(fmt.Sprintf("%v/tcp", port))
+	httpPort := nat.Port("8080/tcp")
+	ports := map[nat.Port][]nat.PortBinding{
+		httpPort: {
+			nat.PortBinding{
+				HostPort: port,
+				HostIP:   "127.0.0.1",
+			},
+		},
+	}
+	return container.HostConfig{PortBindings: ports}, nil
 }
 
-// run command supports only ENV values in from FOO=bar or FOO={{ env:LOCAL_VALUE }}
+func newEnvironmentVariables(f fn.Function, verbose bool) ([]string, error) {
+	// TODO: this has code-smell.  It may not be ideal to have fn.Function
+	// represent Envs as pointers, as this causes the clearly odd situation of
+	// needing to check if an env defined in f is just nil pointers: an invalid
+	// data structure.
+	envs := []string{}
+	for _, env := range f.Envs {
+		if env.Name != nil && env.Value != nil {
+			value, set, err := processEnvValue(*env.Value)
+			if err != nil {
+				return envs, err
+			}
+			if set {
+				envs = append(envs, *env.Name+"="+value)
+			}
+		}
+	}
+	if verbose {
+		envs = append(envs, "VERBOSE=true")
```

Comment:
Hmm - I wonder if we should consider some standard template behavior when this env is set.

Reply:
It would indeed be nice to know that all our various languages' templates treat the verbosity flag in roughly the same way we do here in the client library:  quiet, printing only the data necessary or errors, unless verbose is enabled in which case one gets a play-by-play.
Appropos is [Dave Cheney's "Let's talk about logging"](https://dave.cheney.net/2015/11/05/lets-talk-about-logging)